src/ocaml_monadic_ppx.ml

```diff
@@ -1,113 +1,130 @@
-open Migrate_parsetree;;
-open OCaml_406.Ast;;
-open Ast_mapper;;
-open Asttypes;;
-open Parsetree;;
+open Ppxlib
 
-let ocaml_monadic_mapper =
-  (* We override the expr mapper to catch bind and orzero.  *)
-  { default_mapper with
-    expr = fun mapper outer_expr ->
-      match outer_expr with
-      | [%expr [%bind [%e? expr]]] ->
-        (* Matches "bind"-annotated expressions. *)
-        begin
-          match expr.pexp_desc with
-          | Pexp_let(Nonrecursive, value_bindings, body) ->
-            (* This is a let%bind expression!  It's of the form
-                 let%bind $p1 = $e1 and ... and $pn = $en in $e0
-               and we want it to take the form
-                 bind $e1 (fun $p1 -> ... bind $en (fun $pn -> ...) ...)
-            *)
-            let rec bind_wrap value_bindings' =
-              match value_bindings' with
-              | { pvb_pat = bind_pattern
-                ; pvb_expr = bind_expr
-                ; pvb_attributes = []
-                ; pvb_loc = _bind_loc
-                }::value_bindings'' ->
-                (* Recurse and then wrap the resulting body. *)
-                let body' = bind_wrap value_bindings'' in
-                let cont_function =
-                  [%expr fun [%p bind_pattern] -> [%e body']]
-                    [@metaloc expr.pexp_loc]
-                in
-                [%expr
-                  bind [%e mapper.expr mapper bind_expr] [%e cont_function]]
-                  [@metaloc expr.pexp_loc]
-              | _ ->
-                (* Nothing left to do.  Just return the body. *)
-                mapper.expr mapper body
+let bind_expand ~ctxt expr =
+  let loc = Expansion_context.Extension.extension_point_loc ctxt in
+  match expr.pexp_desc with
+  | Pexp_let (Nonrecursive, value_bindings, body) ->
+      (* This is a let%bind expression!  It's of the form
+                   let%bind $p1 = $e1 and ... and $pn = $en in $e0
+                 and we want it to take the form
+                   bind $e1 (fun $p1 -> ... bind $en (fun $pn -> ...) ...)
+      *)
+      let rec bind_wrap value_bindings' =
+        match value_bindings' with
+        | {
+            pvb_pat = bind_pattern;
+            pvb_expr = bind_expr;
+            pvb_attributes = [];
+            pvb_loc = _bind_loc;
+          }
+          :: value_bindings'' ->
+            (* Recurse and then wrap the resulting body. *)
+            let body' = bind_wrap value_bindings'' in
+            let cont_function =
+              ([%expr fun [%p bind_pattern] -> [%e body']]
+              [@metaloc expr.pexp_loc])
```

Comment:
Are these `[@metaloc ...]` still needed? If I remove them, all the tests still pass

Reply:
Hey, good luck with your exams 💪

> Has `ppx_tools_versioned` been deprecated?

I don't think it's been deprecated, but according to this [blog post](https://discuss.ocaml.org/t/ppx-omp-2-0-0-and-next-steps/6231):

> Other libraries such as ppx_tools or ppx_tools_versioned may continue to be maintained by open source contributors, however they will not be maintained by the OCaml platform and will not receive updates from the platform when new compilers are released. Only ppxlib will receive updates from the platform.

 
Do you have any example of an incorrect `bind` usage that would generate an error without source location (without the `@metaloc` annotation)?

Reply:
> Are these `[@metaloc ...]` still needed? If I remove them, all the tests still pass.

They aren't needed anymore. The reason is the following: In the old metaquot, `[@metaloc ...]` was necessary to attach a location to the node generated by metaquot. The ppxlib metaquot works differently though: when using it, there has to be a variable called `loc` of type `Location.t` in scope. That's the location that will be attached to the node generated. If there is no `loc` in scope when using the ppxlib metaquot, you get a compile error. Since you've defined `loc` in line 4, you're all good. So these `metaloc` attributes are meaningless now and you can remove them.

As a side-note: You're using the location of the extension node, whereas before the location of the payload expression was used. I think they coincide, so that shouldn't make any difference.

Reply:
Thanks for the clarification 😊 I removed the `[@metaloc ...]` and changed the `loc` to `expr.pexp_loc`